crypto/crypto.go

```diff
@@ -0,0 +1,351 @@
+package crypto
+
+import (
+	"bytes"
+	crand "crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/ripemd160"
+
+	"github.com/btcsuite/btcd/btcec"
+	tm_crypto "github.com/tendermint/go-crypto"
+	"golang.org/x/crypto/ed25519"
+)
+
+type CurveType int8
+
+const (
+	CurveTypeSecp256k1 CurveType = iota
+	CurveTypeEd25519
+)
+
+func (k CurveType) String() string {
+	switch k {
+	case CurveTypeSecp256k1:
+		return "secp256k1"
+	case CurveTypeEd25519:
+		return "ed25519"
+	default:
+		return "unknown"
+	}
+}
+
+func CurveTypeFromString(s string) (CurveType, error) {
+	switch s {
+	case "secp256k1":
+		return CurveTypeSecp256k1, nil
+	case "ed25519":
+		return CurveTypeEd25519, nil
+	default:
+		var k CurveType
+		return k, ErrInvalidCurve(s)
+	}
+}
+
+type ErrInvalidCurve string
+
+func (err ErrInvalidCurve) Error() string {
+	return fmt.Sprintf("invalid curve type")
+}
+
+// The types in this file allow us to control serialisation of keys and signatures, as well as the interface
+// exposed regardless of crypto library
+
+type Signer interface {
+	Sign(msg []byte) (Signature, error)
+}
+
+// PublicKey
+type PublicKey struct {
+	CurveType CurveType
+	PublicKey []byte
+}
+
+type PrivateKey struct {
+	CurveType  CurveType
+	PublicKey  []byte
+	PrivateKey []byte
+}
+
+type PublicKeyJSON struct {
```

Comment:
The alternative would be to define:
```go
type KeyBytes []byte

type PrivateKey struct {
	CurveType  CurveType
	PublicKey  KeyBytes
	PrivateKey KeyBytes
}
```

Ane then define `MarshalJSON` and `UnmarshalJSON` for `KeyBytes` and `CurveType` - that we we can get consistent serialisation whereever we use them and we lose the extra struct. I'd also rather stick with "CurveType" and "PublicKey" instead of "type" and "data".... but if we drop the struct we get that anyway. We can scrap the marshla methods for public key and it will work for private key.

Reply:
So with:
// PublicKey
type PublicKey struct {
    CurveType CurveType
    PublicKey KeyBytes
}

func (p PublicKey) MarshalText() ([]byte, error) {
    return json.Marshal(p)
}

func (p *PublicKey) UnmarshalText(text []byte) error {
    return json.Unmarshal(text, p)
}

We have two problems. 1:
MarshalText() gets into an infinite recursion, since json.Marshal() will call MarshallText() on publickey if MarshalJSON() is not defined. The burntsushi toml unmarshaller directly calls MarshalText(), so the only solution I can think of is:
func (p *PublicKey) MarshalJSON() ([]byte, error) {
	// Declare anonymous struct, since json.Marshal(p) would end up calling MarshalText()
	return json.Marshal(struct {
		CurveType CurveType
		PublicKey KeyBytes
	}{p.CurveType, p.PublicKey})
}
Problem 2:
The json unmarshalling code will attempt UnmarshalText() if that is defined (and UnmarshalJSON is not). This will fail since the input is an object, not a string and the decoder will err out with "json: cannot unmarshal object into Go struct field TxInput.PublicKey of type crypto.PublicKey"
Solution: define a non-recursive UnmarshalJSON():


func (p *PublicKey) UnmarshalJSON(text []byte) error {
	// Declare anonymous struct, since json.Marshal(p) would end up calling MarshalText()
	j := struct {
		CurveType CurveType
		PublicKey KeyBytes
	}{}

	err := json.Unmarshal(text, &j)
	if err != nil {
		return err
	}

	p.CurveType = j.CurveType
	p.PublicKey = j.PublicKey

	return nil
}

I could not find any other solutions with are less hideous (we could have a PublicKeyJSON struct like the one we had before.

I suggest that we keep code as proposed -- except renaming the Type/Data to CurveType and PublicKey.